src/Psy/TabCompletion/Matcher/AbstractDefaultParametersMatcher.php

```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace Psy\TabCompletion\Matcher;
+
+abstract class AbstractDefaultParametersMatcher extends AbstractContextAwareMatcher
+{
+    /**
+     * @param \ReflectionParameter[] $reflectionParameters
+     *
+     * @return array
+     */
+    public function getDefaultParameterCompletion(array $reflectionParameters)
+    {
+        $parametersProcessed = array();
+
+        foreach ($reflectionParameters as $parameter) {
+            if (!$parameter->isDefaultValueAvailable()) {
+                return array();
+            }
+
+            $defaultValue = var_export($parameter->getDefaultValue(), true);
```

Comment:
After using this for a bit, it feels like `var_export` isn't 100% right here.  for example, `var_export(array())` is a bit ugly:

```
Psy\Shell::debug($vars = array (
),$boundObject = NULL)
```

Ideally, this would complete as:

```
Psy\Shell::debug($vars = [], $boundObject = NULL)
```

Reply:
This is very true. I suppose I could parse the code using the AST parser and make a custom formatter for shortform everything, but that would be a lot slower than what is there now. Another option would be to throw some regex at this problem, which is probably a little faster but also a lot less easy to modify. Do you have a preference?

Reply:
I just tried mocking this, but it's actually a lot less trivial than I thought initially. I'll table it for now and return to it a little later.

Reply:
Since it's just for default values, a few simple transforms would probably be fine. The most common case where var export is wrong will likely be an empty array? We could start be special casing just that one. And maybe deal with new lines?

Reply:
A filled array is also strange. With multiline strings as default parameters, you get newlines but not ones you want to replace. Then you can have multiline strings inside of arrays as default parameters, that is where I decided I didn't want regex. I spent a little bit getting a prettyprinter to work reliable but failed, but that definitely doesn't mean it's impossible - I just have to mull over what I did wrong.

Reply:
How about this:

```php
function stringify($value) {
  if (is_array($value)) {
    $chunks = [];
    foreach ($value as $key => $val) {
      if ($key !== count($chunks)) {
        return false;
      }

      $chunks[] = stringify($val);
    }

    return '[' . implode(', ', $chunks) . ']';
  }

  // Use json_encode because it's a close enough approximation of what we're
  // looking for with strings, numbers, booleans and null...
  return json_encode($value);
}
```

If we can encode this as a "clean" array, do that, as above. but if `stringify` returns `false`, bail completely and don't complete default parameters.

Using `json_encode` rather than `var_export` solves for `\n` in strings, and special casing all arrays prevents it from generating invalid PHP.

As a bonus, it'll give us `true`, `false`, and `null` rather than the more shouty `TRUE`, `FALSE` and `NULL` that `var_exports` returns.

As an extra bonus, it doesn't use output buffering internally so we can use it wherever we want :)

Reply:
That's an interesting solution. I'll test that out (probably in the morning), thanks!

Reply:
poke :)

Reply:
Apologies, I didn't forget about you :) I got sidetracked by something else. I'll look at this today assuming nothing explodes.

Reply:
No worries. Just wanted to make sure it didn't get dropped.